quality_of_service_demo/src/liveliness.cpp

```diff
@@ -0,0 +1,195 @@
+  printf("%s node_assert_period : How often the Publisher will assert the liveliness of its "
+    "Node (in uint milliseconds). Defaults to 0 (never)\n", OPTION_NODE_ASSERT_PERIOD);
+  printf("%s topic_assert_period : How often the Publisher will assert the liveliness of its "
+    "Publisher (in uint milliseconds). Defaults to 0 (never)\n", OPTION_TOPIC_ASSERT_PERIOD);
+  printf("%s kill_publisher_after : "
+    "Kill the publisher after this amount of time (in uint milliseconds). "
+    "In AUTOMATIC - destroy the whole node. "
+    "In MANUAL_BY_NODE, stop node liveliness assertion. "
+    "In MANUAL_BY_TOPIC, stop topic liveliness assertion. "
+    "Defaults to %lu\n",
+    OPTION_KILL_PUBLISHER_AFTER, DEFAULT_KILL_PUBLISHER_AFTER);
+}
+
+class LivelinessDemo
+{
+public:
+  LivelinessDemo(
+    const std::string & topic,
+    rmw_qos_liveliness_policy_t liveliness_policy_kind,
+    std::chrono::milliseconds liveliness_lease_duration,
+    std::chrono::milliseconds node_assert_period,
+    std::chrono::milliseconds topic_assert_period,
+    std::chrono::milliseconds kill_publisher_after)
+  : liveliness_policy_kind_(liveliness_policy_kind)
+  {
+    rclcpp::QoS qos_profile(10);
+    qos_profile.liveliness(liveliness_policy_kind);
+    qos_profile.liveliness_lease_duration(liveliness_lease_duration);
+
+    rclcpp::SubscriptionOptions subscription_options;
+    subscription_options.event_callbacks.liveliness_callback = std::bind(
+      &LivelinessDemo::print_liveliness_changed_event, this, std::placeholders::_1);
+
+    rclcpp::PublisherOptions publisher_options;
+
+    listener_ = std::make_shared<Listener>(
+      topic, qos_profile, subscription_options);
+    talker_ = std::make_shared<Talker>(
+      topic, qos_profile, publisher_options, 0, node_assert_period, topic_assert_period);
+
+    executor.add_node(listener_);
+    executor.add_node(talker_);
+
+    kill_publisher_timer_ = listener_->create_wall_timer(
+      kill_publisher_after, std::bind(&LivelinessDemo::kill_publisher, this));
+  }
+
+  void
+  spin()
+  {
+    executor.spin();
+  }
+
+private:
+  void
+  print_liveliness_changed_event(rclcpp::QOSLivelinessChangedInfo & event)
+  {
+    printf("Liveliness changed event: \n");
+    printf("  alive_count: %d\n", event.alive_count);
+    printf("  not_alive_count: %d\n", event.not_alive_count);
+    printf("  alive_count_change: %d\n", event.alive_count_change);
+    printf("  not_alive_count_change: %d\n", event.not_alive_count_change);
+  }
+
+  void
+  kill_publisher()
+  {
+    if (!talker_) {
+      return;
+    }
+    switch (liveliness_policy_kind_) {
+      case RMW_QOS_POLICY_LIVELINESS_AUTOMATIC:
+        executor.remove_node(talker_);
+        talker_ = nullptr;
+        break;
+      case RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE:
+        talker_->stop_asserting_liveliness();
+        break;
+      case RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC:
+        talker_->stop_asserting_liveliness();
+        break;
```

Comment:
These two cases could be collapsed:

```c++
      case RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE:
      case RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC:
        talker_->stop_asserting_liveliness();
        break;
```

Reply:
Or maybe this is a copy-paste mistake?

Reply:
It's not a mistake. I was under the impression that switch-fallthrough was generally frowned upon.

Reply:
Ok, you can leave it as-is, but I don't have a problem with intentional fall through.